client/ayon_core/host/interfaces/workfiles.py

```diff
@@ -0,0 +1,1214 @@
+
+        return cls(
+            project_name=project_name,
+            folder_id=folder_id,
+            task_id=task_id,
+            representation_id=repre_entity["id"],
+            filepath=filepath,
+            created_at=created_at.float_timestamp,
+            author=author,
+            available=available,
+            file_size=file_size,
+            file_created=file_created,
+            file_modified=file_modified,
+        )
+
+    def to_data(self) -> dict[str, Any]:
+        """Converts file item to data.
+
+        Returns:
+            dict[str, Any]: Workfile item data.
+
+        """
+        return asdict(self)
+
+    @classmethod
+    def from_data(cls, data: dict[str, Any]) -> "PublishedWorkfileInfo":
+        """Converts data to workfile item.
+
+        Args:
+            data (dict[str, Any]): Workfile item data.
+
+        Returns:
+            PublishedWorkfileInfo: File item.
+
+        """
+        return PublishedWorkfileInfo(**data)
+
+
+class IWorkfileHost:
+    """Implementation requirements to be able to use workfiles utils and tool.
+
+    Some of the methods are pre-implemented as they generally do the same in
+        all host integrations.
+
+    """
+    @abstractmethod
+    def save_workfile(self, dst_path: Optional[str] = None):
+        """Save the currently opened scene.
+
+        Args:
+            dst_path (str): Where the current scene should be saved. Or use
+                the current path if 'None' is passed.
+
+        """
+        pass
+
+    @abstractmethod
+    def open_workfile(self, filepath: str):
+        """Open passed filepath in the host.
+
+        Args:
+            filepath (str): Path to workfile.
+
+        """
+        pass
+
+    @abstractmethod
+    def get_current_workfile(self) -> Optional[str]:
+        """Retrieve a path to current opened file.
+
+        Returns:
+            Optional[str]: Path to the file which is currently opened. None if
+                nothing is opened or the current workfile is unsaved.
+
+        """
+        return None
+
+    def workfile_has_unsaved_changes(self) -> Optional[bool]:
+        """Currently opened scene is saved.
+
+        Not all hosts can know if the current scene is saved because the API
+            of DCC does not support it.
+
+        Returns:
+            Optional[bool]: True if scene is saved and False if has unsaved
+                modifications. None if can't tell if workfiles has
+                modifications.
+
+        """
+        return None
+
+    def get_workfile_extensions(self) -> list[str]:
+        """Extensions that can be used to save the workfile to.
+
+        Notes:
+            Method may not be used if 'list_workfiles' and
+                'list_published_workfiles' are re-implemented with different
+                logic.
+
+        Returns:
+            list[str]: List of extensions that can be used for saving.
+
+        """
+        return []
+
+    def save_workfile_with_context(
```

Comment:
I don't like public functions with some many arguments (and most of them optional). Could we find a better way  to pass these commonly passed things? I am touching again the `folder_entity`, `task_entity` as I've mentioned before. There are several ways to solve it:

- split it into more functions - way more easier to write unit test.
- builder pattern
- group data together in defined structures

Reply:
Do it if you want to do it. I don't know how, and I don't want to spend few day tweaking it to make you satisfied.

Reply:
It is not to satisfy me. I just think that if we are introducing new API, it should be nice to use and test. Maybe @philippe-ynput can decide here. I don't want to be human based "linter" here. But try to write unit tests for functions like these and it is major PITA.

Reply:
I don't understand why it's PITA?

It can require `filepath`, `rootless_path`, `project_entity`, `folder_entity`, `task_entity`, `anatomy`, `workfile_entity_id` and optional `version`, `comment` and `description`.
But that would be PITA to use.